### PR TITLE
OHRM5X-824: Add pim - report custom fields 

### PR DESCRIPTION
**Pull request by SharukaNP**

### CODE REVIEW

File: src/lib/framework/Services.php [hunk 1]
```diff
@@ -258,6 +263,7 @@ final class Services
      */
     public const EMPLOYEE_SALARY_SERVICE = 'pim.employee_salary_service';
 
+
```
Comment (SharukaNP) [inline src/lib/framework/Services.php:266]: Remove empty line

File: src/plugins/orangehrmPimPlugin/Api/PimReportDataAPI.php [hunk 2]
```diff
@@ -52,6 +52,7 @@ protected function getReportGeneratorService(): ReportGeneratorService
     protected function getReport(): EndpointAwareReport
     {
         $reportName = $this->getReportName();
+
```
Comment (SharukaNP) [inline src/plugins/orangehrmPimPlugin/Api/PimReportDataAPI.php:55]: remove empty line

File: src/plugins/orangehrmPimPlugin/Dao/CustomFieldDao.php [hunk 3]
```diff
@@ -64,9 +68,48 @@ public function saveCustomField(CustomField $customField): CustomField
         if (!(strlen((string)$seqNo) <= 10 && $seqNo > 0)) {
```
Comment (SharukaNP) [inline src/plugins/orangehrmPimPlugin/Dao/CustomFieldDao.php:68]: update doc comment for transaction

File: src/plugins/orangehrmPimPlugin/Dao/CustomFieldDao.php [hunk 4]
```diff
@@ -99,16 +142,26 @@ public function getCustomFieldById(int $id): ?CustomField
...
             $q = $this->createQueryBuilder(CustomField::class, 'cf');
-            $q->delete()
-                ->andWhere($q->expr()->in('cf.fieldNum', ':ids'))
-                ->setParameter('ids', $toDeleteIds);
+            $q->delete()->andWhere($q->expr()->in('cf.fieldNum', ':ids'))->setParameter('ids', $toDeleteIds);
             return $q->getQuery()->execute();
         } catch (Exception $e) {
             throw new DaoException($e->getMessage());
         }
     }
```
Comment (SharukaNP) [inline src/plugins/orangehrmPimPlugin/Dao/CustomFieldDao.php:153]: doc comment

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmCorePlugin/Dao/ReportGeneratorDao.php [hunk 5]
```diff
@@ -432,4 +432,55 @@ public function getSkippedSelectedFilterFieldsByReportId(int $reportId): array
...
         return $q->getQuery()->execute();
     }
+
+    /**
+     * @return array
+     */
+    public function getDisplayFieldGroups(): array
+    {
+        $q = $this->createQueryBuilder(DisplayFieldGroup::class, 'displayFieldGroup');
+        $q->select('displayFieldGroup.id,displayFieldGroup.name');
```
[the remark this answers is not in the document]
Comment (SharukaNP) [reply src/plugins/orangehrmCorePlugin/Dao/ReportGeneratorDao.php:445]: Change to object array

File: src/plugins/orangehrmAdminPlugin/Api/EmailConfigurationAPI.php [hunk 6]
```diff
@@ -283,20 +283,23 @@ public function setEmailConfiguration(EmailConfiguration $emailConfiguration)
...
-        );
 
         $password = $this->getRequestParams()->getStringOrNull(
             RequestParams::PARAM_TYPE_BODY,
             self::PARAMETER_SMTP_PASSWORD
         );
 
-        if (!is_null($password)) {
-            $emailConfiguration->setSmtpPassword($password);
+        $emailConfiguration->setSmtpPassword($password);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Api/EmailConfigurationAPI.php:292]: Want to discuss this.

File: src/plugins/orangehrmCorePlugin/test/Dao/ReportGeneratorDaoTest.php [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ * OrangeHRM is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace Dao;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmCorePlugin/test/Dao/ReportGeneratorDaoTest.php:21]:
```suggestion
namespace OrangeHRM\Tests\Core\Dao;
```

File: src/plugins/orangehrmCorePlugin/test/Dao/ReportGeneratorDaoTest.php [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+use OrangeHRM\Config\Config;
+use OrangeHRM\Core\Dao\ReportGeneratorDao;
+use OrangeHRM\Tests\Util\KernelTestCase;
+use OrangeHRM\Tests\Util\TestDataService;
+
+class ReportGeneratorDaoTest extends KernelTestCase
```
Comment (RajithaKumara) [inline src/plugins/orangehrmCorePlugin/test/Dao/ReportGeneratorDaoTest.php:28]: add groups annotations

File: src/plugins/orangehrmPimPlugin/Dao/CustomFieldDao.php [hunk 9]
```diff
@@ -64,9 +69,52 @@ public function saveCustomField(CustomField $customField): CustomField
...
+            $this->rollBackTransaction();
+            throw new TransactionException($exception);
+        }
+    }
+
+    /**
+     * @param CustomField $customField
+     * @return void
+     */
+    public function addCustomFieldToDisplayField(CustomField $customField): void
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPimPlugin/Dao/CustomFieldDao.php:88]:
```suggestion
    private function addCustomFieldToDisplayField(CustomField $customField): void
```

File: src/plugins/orangehrmPimPlugin/Dao/CustomFieldDao.php [hunk 10]
```diff
@@ -94,18 +142,35 @@ public function getCustomFieldById(int $id): ?CustomField
...
+            $this->rollBackTransaction();
+            throw new TransactionException($exception);
+        }
+    }
+
+    /**
+     * @param array $toDeleteCustomFieldIds
+     * @return void
+     */
+    public function deleteDisplayFieldsOfCustomFields(array $toDeleteCustomFieldIds): void
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPimPlugin/Dao/CustomFieldDao.php:166]:
```suggestion
    private function deleteDisplayFieldsOfCustomFields(array $toDeleteCustomFieldIds): void
```

File: src/plugins/orangehrmPimPlugin/Dao/CustomFieldDao.php [hunk 11]
```diff
@@ -94,18 +142,35 @@ public function getCustomFieldById(int $id): ?CustomField
...
+    {
+        foreach ($toDeleteCustomFieldIds as $toDeleteCustomFieldId) {
+            $customField = $this->getCustomFieldById($toDeleteCustomFieldId);
+            $displayFieldName = 'hs_hr_employee.custom' . $customField->getFieldNum();
+            $q = $this->createQueryBuilder(DisplayField::class, 'displayField');
+            $q->delete()->andWhere('displayField.name = :displayFieldName')->setParameter('displayFieldName', $displayFieldName);
+            $q->getQuery()->execute();
         }
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPimPlugin/Dao/CustomFieldDao.php:174]:
```php
        if (empty($toDeleteCustomFieldIds)) {
            return;
        }
        $displayFieldNames = [];
        foreach ($toDeleteCustomFieldIds as $toDeleteCustomFieldId) {
            $customField = $this->getCustomFieldById($toDeleteCustomFieldId);
            $displayFieldNames[] = 'hs_hr_employee.custom' . $customField->getFieldNum();
        }
        $q = $this->createQueryBuilder(DisplayField::class, 'displayField');
        $q->delete()
            ->andWhere($q->expr()->in('displayField.name', ':displayFieldNames'))
            ->setParameter('displayFieldNames', $displayFieldNames);
        $q->getQuery()->execute();
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**